pydantic/_internal/_fields.py

```diff
@@ -246,10 +246,18 @@ def collect_fields(  # noqa: C901
             # 1. To match the behaviour of annotation-only fields
             # 2. To avoid false positives in the NameError check above
             try:
-                delattr(cls, ann_name)
                 if cls.__pydantic_generic_parameters__:  # model can be parametrized
                     assert cls.__pydantic_generic_defaults__ is not None
                     cls.__pydantic_generic_defaults__[ann_name] = default
+                delattr(cls, ann_name)
+                if (
+                    is_dataclass
+                    and isinstance(default, (dataclasses.Field, FieldInfo))
+                    and default.default not in (Undefined, dataclasses.MISSING)
+                ):
+                    # for dataclasses we only delete the attribute if it is an instance of
+                    # field, e.g. `a: int = 1` should not get deleted
+                    setattr(cls, ann_name, default.default)
```

Comment:
I think the comment above this line has become a bit confusing now that it is right above where you setattr, with all of the above conditions. Not a big deal if you don't want to make more changes